### PR TITLE
Returns an UserAgent struct instead of multiple variables

### PR DESCRIPTION
**Pull request by c2h5oh**
It makes it easier to use - you don't have to check which return is what, just use correct element of the struct. 

It also makes it easier to add new elements to returned value without breaking backwards compatibility, which I intend to do in later pull requests starting with minor and patch parts of version.

### CODE REVIEW

**Comment by arussellsaw** [comment]
Hey! Thanks for the PR, this is a pretty hefty public api change so i'll wait for @jacobpierce or @ryanslade to come online and give the final approval, but i may as well offer comment on something i've noticed.

File: uasurfer_test.go [hunk 1]
```diff
@@ -7,710 +7,903 @@ import (
...
-	browserName    BrowserName
+	UA string
+	UserAgent
+	/*browserName    BrowserName
 	browserVersion int
 	Platform       Platform
 	osName         OSName
 	osVersion      int
-	deviceType     DeviceType
+	deviceType     DeviceType*/
```
Comment (arussellsaw) [inline uasurfer_test.go]: could you remove this commented code?

**Comment by iand** [comment]
:+1:

**Comment by c2h5oh** [comment]
Ah yes, I kept that for reference while coding - cleaned up.

**Comment by jacobpierce** [comment]
@c2h5oh I will go through these PRs in the next day or two :jack_o_lantern:

**Comment by c2h5oh** [comment]
@jacobpierce Great! 
This is not the last of it, so if you could provide me with some test data I asked about here https://github.com/avct/uasurfer/issues/18 that would be great. I could use our own logs, but despite large volume I feel the user agent sample would not be diverse enough.

**Comment by c2h5oh** [comment]
bump?

**Comment by c2h5oh** [comment]
weekly bump

If you don't like any of the proposed changes (either in this pull request or the other two) please voice them, even briefly, so I can work on fixing them (or I know that I have to maintain a fork that fits our needs) - I prefer a rejected pull request to a one hanging in limbo.

**Comment by jacobpierce** [comment]
@c2h5oh Apologies for taking so long. I really like the direction of this PR, and like seeing structs returned. Looking at performance changes, I've rerun the benchmarks about a dozen times and there wasn't much variation but it did show a 40-60% performance/resource downgrade with the PR:

**Current uasurfer master:**

```
BenchmarkAgentSurfer-16                 200000         10498 ns/op         265 B/op           3 allocs/op
BenchmarkEvalSystem-16                  300000          3760 ns/op           0 B/op           0 allocs/op
BenchmarkEvalBrowserName-16            2000000           604 ns/op           0 B/op           0 allocs/op
BenchmarkEvalBrowserVersion-16          500000          2362 ns/op          15 B/op           0 allocs/op
BenchmarkEvalDevice-16                 2000000           624 ns/op           0 B/op           0 allocs/op
BenchmarkParseChromeMac-16              300000          7163 ns/op         288 B/op           4 allocs/op
BenchmarkParseChromeWin-16              200000          6234 ns/op         240 B/op           3 allocs/op
BenchmarkParseChromeAndroid-16          200000         10624 ns/op         320 B/op           4 allocs/op
BenchmarkParseSafariMac-16              300000          6570 ns/op         288 B/op           4 allocs/op
BenchmarkParseSafariiPad-16             200000          6788 ns/op         288 B/op           4 allocs/op
PASS
ok      code.avct.io/uasurfer    17.671s
```

**PR 'structured' branch:**

```
BenchmarkAgentSurfer-16                 100000         10974 ns/op         330 B/op           5 allocs/op
BenchmarkEvalSystem-16                  300000          3530 ns/op           1 B/op           0 allocs/op
BenchmarkEvalBrowserName-16            3000000           592 ns/op           0 B/op           0 allocs/op
BenchmarkEvalBrowserVersion-16         1000000          2108 ns/op          22 B/op           0 allocs/op
BenchmarkEvalDevice-16                 2000000           586 ns/op           0 B/op           0 allocs/op
BenchmarkParseChromeMac-16              300000          9593 ns/op         384 B/op           6 allocs/op
BenchmarkParseChromeWin-16              200000          8754 ns/op         352 B/op           6 allocs/op
BenchmarkParseChromeAndroid-16          100000         14682 ns/op         432 B/op           6 allocs/op
BenchmarkParseSafariMac-16              200000          9857 ns/op         368 B/op           6 allocs/op
BenchmarkParseSafariiPad-16             200000         11142 ns/op         368 B/op           6 allocs/op
PASS
ok      github.com/c2h5oh/uasurfer    19.402s
```

It would be optimal to return a struct without losing significant performance or raising resource requirements.

I'll add something like this to the readme, but wanted to mention: The existing priorities behind uasurfer are to favor performance, then accuracy and then resolution (e.g. minor version). Currently it's used to evaluate billions of strings per day with low overhead and high turnaround. This is almost opposite to great projects like http://github.com/faisalman/ua-parser-js that can be as slow as they like in favor of accuracy/resolution.

What are your goals for this branch and onward? Ease of use and greater parsing resolution? Perhaps we can work on a structure that can support multiple use cases? Possibly two modes of operation that share tests and logic, one high performance and one high resolution.

**Comment by c2h5oh** [comment]
Re: goals:
A compromise between the three, favoring speed, for example:

Resolution:
For Android devices major version is insufficient for pretty much anything but showing a metric.
Android doesn't use semantic versioning, so a minor version often means a breaking change. A different API version. Different TLS version supported and a lot of other things. 
Adding minor version support is most likely quite cheap performance wise - one extra cycle in regex state machine, some extra memory pressure, but it is bound to increase memory pressure. Not free, but justifiable considering the wide range of users this is a deal breaker for.

Accuracy: 
99%+ for everything but OS version and 99.9% for 4 out of 6 (with 1 more being close) makes 99.9 or even 99.99 look achievable. I'd be willing to sacrifice 10% of performance for that. Maybe it can be free. If you can share test data I can tinker and see what results I can get.


Re: performance degradation
System, Browser Name/Version and device type are marginally faster or within a margin of error. Os name/version ones seem to have suffered a 50% performance hit - I'll look into that.

**Comment by c2h5oh** [comment]
I can't replicate the performance degradation you've noticed - based on 10 runs for each version - current master and this pull request

```
# benchcmp -best -changed a27c4e4249e7361e683e87b125f113215f705023.txt 11fce5342a1ac5d653f9c509a4dd167349165d53.txt
benchmark                         old ns/op     new ns/op     delta
BenchmarkAgentSurfer-8            5263          5254          -0.17%
BenchmarkEvalSystem-8             2502          2461          -1.64%
BenchmarkEvalBrowserName-8        543           506           -6.81%
BenchmarkEvalBrowserVersion-8     1180          1185          +0.42%
BenchmarkEvalDevice-8             522           499           -4.41%
BenchmarkParseChromeMac-8         3216          3256          +1.24%
BenchmarkParseChromeWin-8         2956          2928          -0.95%
BenchmarkParseChromeAndroid-8     5887          5862          -0.42%
BenchmarkParseSafariMac-8         3540          3541          +0.03%
BenchmarkParseSafariiPad-8        4258          4289          +0.73%
```

it's most likely because c2h5oh/uasurfer includes more changes that are not part of this pull request

**Comment by jacobpierce** [comment]
Indeed, didn't run tests on the correct branch, thanks for double checking -- looks good. Really pleased with this PR.

**Pull request merged**